File: windows/MoDi.Desktop/Links/WifiDirect/WifiDirectLink.cs
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
using System;
using System.Threading;
using System.Threading.Tasks;
using MoDi.Core;
using MoDi.Protocol;
using MoDi.Desktop.Core.Session;
using MoDi.Core.Factory;
using MoDi.Core.Infrastructure;
using MoDi.Desktop.Diagnostics;

namespace MoDi.Desktop.Links;

/// <summary>
/// WifiDirectLink — WiFi Direct P2P 链路（完整实现）
///
/// 职责：P2P 发现/连接 + QR 码 + 主动向 Android GO 发 HELLO 握手。
/// 与 LAN / 蓝牙 / USB 完全解耦。
/// </summary>
public sealed class WifiDirectLink : ILink
{
    private const string Tag = "WifiDirectLink";

    // ── 兼容别名；稳定值由 TransportIdentity 统一定义 ──
    public const int AudioPort = TransportIdentity.AudioPort;
    public const int HandshakePort = TransportIdentity.HandshakePort;

    // ── 核心模块 ──
    private WifiDirectP2pHelper? _p2pHelper;
    private readonly ConnectionStateManager _stateManager;
    private readonly Func<int, bool> _onHandshakeRoute;
    private CancellationTokenSource? _helloCts;  // 握手任务取消令牌

    // ── 事件（LinkManager / UI 订阅） ──
    public Action<string>? OnP2pStatusChanged;
    public Action<string?, string?>? OnQrChanged;
    public Action<bool>? OnP2pProgressVisible;
    public Action<bool, double>? OnP2pProgress;
    public Action<bool>? OnP2pActiveChanged;
    public Action<Guid>? OnSessionStarted;
    public Action<Guid>? OnSessionEnded;

    private bool _sessionActive;
    private Guid? _sessionId;

    public LinkState State { get; private set; } = LinkState.Idle;
    public bool IsActive => State != LinkState.Idle;

    public WifiDirectLink(
        ConnectionStateManager stateManager,
        Func<int, bool> onHandshakeRoute)
    {
        _stateManager = stateManager;
        _onHandshakeRoute = onHandshakeRoute;
    }

    // ── ILink 实现 ──

    /// <summary>启动 P2P 链路（创建 P2pHelper + 生成 QR 码 + 等待手机扫码连接）</summary>
    public async Task<bool> ConnectAsync()
    {
        if (State != LinkState.Idle) return true;
        State = LinkState.Listening;

        _p2pHelper = new WifiDirectP2pHelper();
        _p2pHelper.OnStatusChanged += msg =>
        {
            OnP2pProgressVisible?.Invoke(true);
            OnP2pStatusChanged?.Invoke(msg);
        };

        var qrContent = QrCodeHelper.BuildQrPayload(_p2pHelper.DeviceName, _p2pHelper.Token);
        OnQrChanged?.Invoke(qrContent, _p2pHelper.DeviceName);

        _helloCts = new CancellationTokenSource();
        var helloToken = _helloCts.Token;
        _p2pHelper.OnConnected += () => GlobalExceptionBoundary.Observe(
            Task.Run(() => SendHelloToAndroidGo(helloToken)),
            "WifiDirect.SendHello");
        _p2pHelper.OnDisconnected += EndSession;

        OnP2pActiveChanged?.Invoke(true);

        await _p2pHelper.StartAsync();
        return true;
    }

    /// <summary>停止 P2P 链路（取消握手 + 停止发现 + 清理 QR）</summary>
    public async Task DisconnectAsync()
    {
        if (_p2pHelper == null) return;

        // 先取消正在进行的握手任务
        _helloCts?.Cancel();
        _helloCts?.Dispose();
        _helloCts = null;

        await _p2pHelper.StopAsync();
        EndSession();
        _p2pHelper.Dispose();
        _p2pHelper = null;

        OnP2pActiveChanged?.Invoke(false);
        OnQrChanged?.Invoke(null, null);
        OnP2pProgressVisible?.Invoke(false);
        OnP2pStatusChanged?.Invoke("");
        OnP2pProgress?.Invoke(true, 0);
        State = LinkState.Idle;
    }

    // ── P2P 握手（主动向 Android GO 发 HELLO） ──
    // 流程：P2P 连接建立 → 等待 3s（Android 就绪）→ 发 HELLO(token) → 等 ACK → 设置路由

    private const int HelloInitialDelayMs = 3_000;   // 等 Android 端 waitForHello 就绪
    private const int HelloMaxAttempts = 6;           // 重试次数（覆盖 Android 60s 监听窗口）
    private const int HelloTimeoutMs = 3_000;         // 每次等待 ACK 超时
    private const int HelloRetryDelayMs = 2_000;      // 重试间隔

    /// <summary>向 Android GO 发送 HELLO 并等待 ACK（最多重试 6 次，覆盖 Android 60s 监听窗口）</summary>
    private async Task SendHelloToAndroidGo(CancellationToken ct)
    {
        ITransport? transport = null;
        try
        {
            var goIp = WifiDirectP2pHelper.GoIp;
            var token = _p2pHelper?.Token ?? "";

            // 等待 P2P 网络稳定 + Android 端 waitForHello 开始监听
            Log.I(Tag, $"P2P connected, waiting {HelloInitialDelayMs}ms before HELLO...");
            OnP2pStatusChanged?.Invoke("P2P 已连接，等待手机端就绪...");
            await Task.Delay(HelloInitialDelayMs, ct);

            Log.I(Tag, $"Sending HELLO to Android GO: {goIp}:{TransportIdentity.HandshakePort}");
            transport = PlatformFactory.CreateTransport(
                TransportType.Udp,
                goIp,
                TransportIdentity.HandshakePort);
            var protocol = PlatformFactory.CreateProtocol();
            await transport.ConnectAsync();

            var sessionId = Guid.NewGuid();
            var payload = HelloSessionPayload.Encode(0, token, sessionId);

            var packet = new Packet
            {
                Type = PacketType.Hello,
                LinkType = LinkType.WifiDirect,
                Sequence = 0,
                Payload = payload
            };
            var encoded = protocol.Encode(packet);

            for (int i = 0; i < HelloMaxAttempts; i++)
            {
                ct.ThrowIfCancellationRequested();
                await transport.SendAsync(encoded);
                Log.I(Tag, $"HELLO sent to {goIp}:{TransportIdentity.HandshakePort} (attempt {i + 1}/{HelloMaxAttempts})");
                OnP2pStatusChanged?.Invoke($"正在握手...（{i + 1}/{HelloMaxAttempts}）");

                try
                {
                    var reply = await WaitForPacketAsync(transport, HelloTimeoutMs);
                    if (reply != null)
                    {
                        var decoded = protocol.Decode(reply.Value.Span);
                        if (decoded.HasValue &&
                            decoded.Value.Type == PacketType.HelloAck &&
                            HelloSessionPayload.MatchesAck(decoded.Value.Payload, sessionId))
                        {
                            HelloSessionPayload.TryDecode(
                                decoded.Value.Payload,
                                tokenRequired: false,
                                out var ackIdentity);
                            var route = ackIdentity.Route;

                            Log.I(Tag, $"HELLO_ACK received! P2P handshake OK, route={route}");
                            _stateManager.BeginConnecting();

                            if (!_sessionActive)
                            {
                                _sessionActive = true;
                                _sessionId = sessionId;
                                OnSessionStarted?.Invoke(sessionId);
                            }
                            _onHandshakeRoute(route);

                            // 配对持久化：握手成功即写入，后续冷启动免扫码
                            var paired = PairedDeviceStore.GetOrCreate();
                            paired.LastConnected = DateTime.Now;
                            PairedDeviceStore.Save(paired);

                            OnP2pStatusChanged?.Invoke($"P2P 握手成功 ✓ local={_p2pHelper?.LocalIp} go={goIp} route={route}");
                            OnP2pProgress?.Invoke(false, 100);
                            OnQrChanged?.Invoke(null, null);
                            _stateManager.Update(ConnectionState.Connected);
                            return;
                        }
                    }
                }
                catch (OperationCanceledException) when (!ct.IsCancellationRequested) { /* timeout, retry */ }

                // 重试前等待
                if (i < HelloMaxAttempts - 1)
                    await Task.Delay(HelloRetryDelayMs, ct);
            }

            Log.W(Tag, $"P2P handshake failed: no HELLO_ACK after {HelloMaxAttempts} attempts");
            OnP2pStatusChanged?.Invoke("P2P 握手失败（手机未响应），等待重连...");
        }
        catch (OperationCanceledException)
        {
            Log.I(Tag, "HELLO task cancelled (P2P stopped)");
        }
        catch (Exception ex)
        {
            Log.E(Tag, $"SendHelloToAndroidGo error: {ex.Message}");
        }
        finally
        {
            if (transport != null) await transport.DisconnectAsync();
        }
    }

    /// <summary>等待收到一个数据包（带超时，用于等待 HELLO_ACK）</summary>
    private static async Task<ReadOnlyMemory<byte>?> WaitForPacketAsync(ITransport transport, int timeoutMs)
    {
        var tcs = new TaskCompletionSource<ReadOnlyMemory<byte>>();
        using var cts = new CancellationTokenSource(timeoutMs);
        using var reg = cts.Token.Register(() => tcs.TrySetCanceled());
        Action<ReadOnlyMemory<byte>> handler = data => tcs.TrySetResult(data);
        transport.PacketReceived += handler;
        try { return await tcs.Task; }
        catch (OperationCanceledException) { return null; }
        finally { transport.PacketReceived -= handler; }
    }

    public void Dispose()
    {
        EndSession();
        _p2pHelper?.Dispose();
    }

    private void EndSession()
    {
        if (!_sessionActive) return;
        _sessionActive = false;
        var endedSessionId = _sessionId;
        _sessionId = null;
        if (endedSessionId.HasValue) OnSessionEnded?.Invoke(endedSessionId.Value);
    }
}

File: android/app/src/main/java/com/modi/connect/links/wifidirect/WifiDirectLink.kt
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.modi.connect.links.wifidirect

import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjection
import com.modi.connect.ConnectionState
import com.modi.connect.ConnectionStateManager
import com.modi.connect.StreamingService
import com.modi.connect.audio.AudioPipeline
import com.modi.connect.core.TransportIdentity
import com.modi.connect.links.ILink
import com.modi.connect.links.LinkState
import com.modi.connect.links.LinkManager
import com.modi.connect.links.LinkParams
import com.modi.connect.net.HandshakeManager
import com.modi.protocol.LinkType
import com.modi.connect.net.P2pPairStore
import com.modi.connect.net.WifiDirectManager
import com.modi.connect.session.DisconnectReason
import com.modi.connect.session.SessionControlMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.UUID

/**
 * WifiDirectLink — WiFi Direct P2P 链路（完整实现）
 *
 * 职责：扫码 → createGroup → 等待 Windows HELLO → 推流。
 * 与 LAN / 蓝牙 / USB 完全解耦。
 */
class WifiDirectLink(
    private val context: Context,
    private val pipe: AudioPipeline,
    private val stateManager: ConnectionStateManager
) : ILink {

    companion object {
        const val AUDIO_PORT = TransportIdentity.AUDIO_PORT
        const val HANDSHAKE_PORT = TransportIdentity.HANDSHAKE_PORT
        const val HANDSHAKE_TIMEOUT_MS = 3000L
        const val HANDSHAKE_MAX_ATTEMPTS = 3
        const val IP_POLL_INTERVAL_MS = 500L
        const val IP_POLL_MAX_RETRIES = 16
    }

    // ── 子模块 ──
    val wifiDirectManager = WifiDirectManager(context)
    private val connectMutex = Mutex()  // 防止并发 connect（扫码过快时两次调用冲突）

    // ── ILink 状态 ──
    @Volatile override var isStreaming = false
        private set
    override val state: LinkState
        get() = if (isStreaming) LinkState.STREAMING else LinkState.IDLE
    override var onStatusChanged: ((String) -> Unit)? = null
    override var onStreamingChanged: ((Boolean) -> Unit)? = null
    @Volatile override var sessionId: UUID? = null
        private set

    // ── P2P 特有状态（链路自治，不依赖外部） ──
    @Volatile var p2pTargetIp: String? = null
        private set
    @Volatile var currentRoute: Int = 0
    private var p2pLocalIp: String? = null       // Android P2P 接口 IP（GO IP）
    private var lastRemoteIp: String? = null     // 上次握手的 Windows IP（重连用）

    // ── ILink 实现 ──

    /**
     * 连接 P2P 链路（用户扫码触发）
     * 流程：获取 token → createGroup → 等待 Windows 连接 → 握手 → 推流
     * 重连策略：已知 Windows IP 时主动发 HELLO，否则被动等待
     */
    override suspend fun connect(params: LinkParams): Boolean = connectMutex.withLock {
        // token 来源：扫码传入 或 已配对存储（冷启动免扫码）
        val token = params.token
            ?: P2pPairStore.load(context)?.token
            ?: return@withLock false
        currentRoute = params.route

        onStatusChanged?.invoke("正在创建 P2P Group...")
        val goIp = wifiDirectManager.createGroupAndWaitForClient()
        if (goIp == null) {
            onStatusChanged?.invoke("P2P Group 创建失败")
            stateManager.update(ConnectionState.ERROR)
            return@withLock false
        }
        p2pLocalIp = goIp

        stateManager.update(ConnectionState.CONNECTING)

        // 重连策略：已知 Windows IP 时主动发 HELLO，否则被动等待
        var establishedSessionId: UUID? = null
        val handshakeOk = if (lastRemoteIp != null) {
            onStatusChanged?.invoke("P2P 重连中，主动握手...")
            val newSessionId = UUID.randomUUID()
            withContext(Dispatchers.IO) {
                HandshakeManager.handshake(
                    lastRemoteIp!!,
                    params.route,
                    token,
                    LinkType.WIFI_DIRECT,
                    p2pLocalIp,
                    newSessionId
                ).also { if (it) establishedSessionId = newSessionId }
            }
        } else {
            onStatusChanged?.invoke("P2P GO 就绪 ($goIp)，等待电脑连接...")
            val remoteIp = withContext(Dispatchers.IO) {
                P2pHandshakeServer.waitForHello(token, params.route)
            }
            if (remoteIp != null) {
                lastRemoteIp = remoteIp.remoteIp
                establishedSessionId = remoteIp.sessionId
                true
            } else false
        }
        if (!handshakeOk) {
            onStatusChanged?.invoke("P2P 握手失败（电脑未响应）")
            stateManager.update(ConnectionState.ERROR)
            return@withLock false
        }

        stateManager.update(ConnectionState.CONNECTED)
        sessionId = establishedSessionId
        onStatusChanged?.invoke("P2P 握手成功 ✓ 准备推流")

        // 配对持久化：握手成功即写入，后续冷启动免扫码
        P2pPairStore.save(context, token, params.deviceName ?: "")

        val winP2pIp = lastRemoteIp ?: goIp
        p2pTargetIp = winP2pIp
        val capMode = LinkManager.routeToCapture(params.route)

        pipe.onFirstFrame = { stateManager.update(ConnectionState.STREAMING) }
        val ok = withContext(Dispatchers.IO) {
            pipe.currentLinkType = LinkType.WIFI_DIRECT
            pipe.startStreaming(capMode, params.proj, context, winP2pIp, localBindAddress = p2pLocalIp)
        }

        if (ok) {
            isStreaming = true
            onStatusChanged?.invoke("P2P 推流中：路线${params.route + 1}")
            onStreamingChanged?.invoke(true)
            context.startForegroundService(Intent(context, StreamingService::class.java))
        } else {
            onStatusChanged?.invoke("P2P 启动推流失败")
            stateManager.update(ConnectionState.ERROR)
        }
        ok
    }

    /** 推流中热切路线：切换采集模式 + 通过 UDP 发送 ROUTE 包到 Windows */
    override suspend fun sendRouteUpdate(route: Int, proj: MediaProjection?): Boolean {
        if (!isStreaming) { currentRoute = route; return true }
        currentRoute = route

        val capMode = LinkManager.routeToCapture(route)
        val ok = withContext(Dispatchers.IO) { pipe.switchMode(capMode, proj, context) }
        if (!ok) { onStatusChanged?.invoke("需先授权系统音频"); return false }

        val targetIp = p2pTargetIp ?: return false
        withContext(Dispatchers.IO) { HandshakeManager.sendRouteUpdate(targetIp, route, LinkType.WIFI_DIRECT, p2pLocalIp) }
        return true
    }

    override suspend fun sendDisconnectRequest(targetLink: Byte, reason: DisconnectReason): Boolean {
        val currentId = sessionId ?: return false
        val targetIp = p2pTargetIp ?: return false
        val message = SessionControlMessage.request(currentId, LinkType.WIFI_DIRECT, targetLink, reason)
        return withContext(Dispatchers.IO) {
            HandshakeManager.sendSessionControl(targetIp, message, p2pLocalIp)
        }
    }

    /** 断开 P2P 链路：停止推流 + 销毁 P2P Group + 状态回退 */
    override fun disconnect() {
        context.stopService(Intent(context, StreamingService::class.java))
        pipe.stopStreaming()
        wifiDirectManager.disconnect()
        isStreaming = false
        p2pTargetIp = null
        p2pLocalIp = null
        sessionId = null
        stateManager.update(ConnectionState.DISCONNECTED)
        onStatusChanged?.invoke("已停止")
        onStreamingChanged?.invoke(false)
    }

    /** 仅在显式重新配对时清除旧电脑，普通断线保留该地址用于快速重连。 */
    fun forgetRemotePeer() {
        lastRemoteIp = null
    }

}
